plugins/module_utils/cloud_py.py

```diff
@@ -0,0 +1,185 @@
+# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+# USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import time
+import functools
+import random
+
+
+def exponential_value_generator(delay, backoff, max_delay=None, jitter=False):
+    """Generates sleep value based on the exponential back-off algorithm.
+    Args:
+        delay (int or float): initial delay.
+        backoff (int or float): backoff multiplier e.g. value of 2 will  double the delay each retry.
+        max_delay (int or None): maximum amount of time to wait between retries.
+        jitter (bool): if set to true, add jitter to the generate value.
+    Yields:
+        float: successive sleep value.
+    """
+    current_delay = delay
+    while True:
+        sleep_time = current_delay
+        return_value = sleep_time if max_delay is None else min(sleep_time, max_delay)
+        if jitter:
+            return_value = random.uniform(0.0, return_value)
+        yield return_value
+        current_delay = current_delay * backoff
+
+
+def retry_func(func, sleep_time_generator, retries, catch_extra_error_codes, found_f, status_code_from_except_f):
+    counter = 0
+    for sleep_time in sleep_time_generator:
+        try:
+            return func()
+        except Exception as exc:
+            counter += 1
+            if counter == retries:
+                raise
+            status_code = status_code_from_except_f(exc)
+            if found_f(status_code, catch_extra_error_codes):
+                time.sleep(sleep_time)
+            else:
+                raise
+
+
+class CloudRetry:
+    """
+    The base class to be used by other cloud providers to provide a backoff/retry decorator based on status codes.
+    """
+
+    @staticmethod
+    def status_code_from_exception(error):
+        if hasattr(error, "status"):
+            return getattr(error, "status")
+        return None
+
+    @staticmethod
+    def found(response_code, catch_extra_error_codes=None):
+        def _is_iterable():
+            try:
+                it = iter(catch_extra_error_codes)
+            except TypeError:
+                # not iterable
+                return False
+            else:
+                # iterable
+                return True
+        return _is_iterable() and response_code in catch_extra_error_codes
+
+    @classmethod
+    def exponential_backoff(cls, retries=10, delay=3, backoff=1.1, max_delay=60, catch_extra_error_codes=None):
```

Comment:
```suggestion
    def exponential_backoff(cls, retries=10, delay=3, backoff=2, max_delay=60, catch_extra_error_codes=None):
```
My fault, there's actually a bad docstring value in the original and I didn't catch it when I did the spec.  The default backoff for this method should be 2.  Please update the docstring here as well!